### ISSUE TITLE
VEP 100 extremely slow in an high performant machine

### ISSUE THREAD

**Issue by cccnrc** [Ensembl/ensembl-vep#763]
I installed VEP 100 and optional perl modules following [website instructions](http://useast.ensembl.org/info/docs/tools/vep/script/vep_download.html). 

I am working on a 

> Ubuntu - DELL Latitude 5410 - 32GB RAM - i7 10th generation

A also tabixed all caches with 
`perl convert_cache.pl -species all -version all --remove`

that outputted at first time:

> 2020-05-18 12:52:06 - Processing homo_sapiens_merged
> 2020-05-18 12:52:06 - Processing version 100_GRCh37
> 2020-05-18 12:52:06 - INFO: BAM-edited cache detected, enabling --use_transcript_ref; use --use_given_ref to override this
> 2020-05-18 12:52:06 - No unprocessed types remaining, skipping
> 2020-05-18 12:52:06 - Processing homo_sapiens
> 2020-05-18 12:52:06 - Processing version 100_GRCh37
> 2020-05-18 12:52:06 - No unprocessed types remaining, skipping
> 2020-05-18 12:52:06 - Processing version 100_GRCh38
> 2020-05-18 12:52:06 - No unprocessed types remaining, skipping
> 2020-05-18 12:52:06 - Processing fasta
> 2020-05-18 12:52:06 - All done!

Now I am trying to see performance and they are awful! Based on `examples/homo_sapiens_GRCh38.vcf` that are only 173 variants long it takes >7 min! My command:

`time perl vep -v -i examples/homo_sapiens_GRCh38.vcf --format vcf --cache --dir_cache /home/enrico/.vep --cache_version 100 --fasta /home/enrico/.vep/fasta/Homo_sapiens.GRCh38.dna.toplevel.fa.gz --vcf --offline --assembly GRCh38 --af_gnomad -o examples/forked.vcf --no_stats --hgvs`
takes:

> real	7m6.930s
> user	7m6.588s
> sys	0m0.316s

and is completely unaffected by adding `--fork 4` parameter:

> real	7m26.177s
> user	7m25.316s
> sys	0m0.781s

Does anybody got a clue on why is it so slow??

Thank you very much in advance

**Comment by aparton** [comment]
Hi @cccnrc,

That is remarkably slow - I've attempted to run the same query on my local machine and it runs in 4 seconds, so there's clearly an issue here.

Our caches should all be tabix converted by default, so converting them should no longer be necessary. It may be worth redownloading all of your cache files with the installer.

To help us diagnose this, could you please run the following and send me the output:

1) `tabix --version`
2) `perl -e 'use Bio::DB::HTS::Tabix`
3) `sum /home/enrico/.vep/homo_sapiens/100_GRCh38/22/all_vars.gz`
4) Could you please try unzipping your fasta file and rerunning your query?

Kind Regards,
Andrew

**Comment by cccnrc** [comment]
Hi @aparton thanks for your reply!

`tabix --version`

> tabix (htslib) 1.10.2
> Copyright (C) 2019 Genome Research Ltd.

`perl -e 'use Bio::DB::HTS::Tabix'`
I got no output

`sum /home/enrico/.vep/homo_sapiens/100_GRCh38/22/all_vars.gz`

> 36085 120767

(what's the last command for?)

With unzipped fasta i spare just 2 minutes:
`time perl vep -v -i examples/homo_sapiens_GRCh38.vcf --format vcf --cache --dir_cache /home/enrico/.vep --cache_version 100 --fasta /home/enrico/.vep/fasta/Homo_sapiens.GRCh38.dna.toplevel.fa --vcf --offline --assembly GRCh38 --af_gnomad -o examples/forked.vcf --no_stats --hgvs --fork 4`

> real	5m2.604s
> user	4m51.470s
> sys	0m11.121s

**Comment by cccnrc** [comment]
Hi @aparton  any update on this?

**Comment by aparton** [comment]
Hi @cccnrc,

Thank you for running those commands. Unfortunately, there's nothing out of the ordinary in your results, and I'm still none-the-wiser as to why you're seeing VEP run so slowly.

1) Could you please try and run `time perl vep -id '1 1234567 . A G' --cache --dir /home/enrico/.vep --cache_version 100 --fasta /home/enrico/.vep/fasta/Homo_sapiens.GRCh38.dna.toplevel.fa --offline --assembly GRCh38`? I just want to make sure that it is the core vep functionality that's causing this issue, rather than any of the flags.

2) If your machine has an internet connection, could you also please try running your command with --database instead of --cache so that we can compare times there as well?

3) If you're willing to try profiling your run, we could try running NYTProf and the results should tell us where the slowdown is happening. https://metacpan.org/pod/Devel::NYTProf

In short, if you install NYTProf with `cpanm Devel::NYTProf` and then your command with `perl -d:NYTProf vep -i examples/...`, then we should be able to see from the output exactly which lines are causing your slowdown.

Kind Regards,
Andrew

**Comment by cccnrc** [comment]
Hi @aparton , thank you for that.

1. 
`
time perl vep -id '1 1234567 . A G' --cache --dir /home/enrico/.vep --cache_version 100 --fasta /home/enrico/.vep/fasta/Homo_sapiens.GRCh38.dna.toplevel.fa --offline --assembly GRCh38 --force_overwrite
`

> real	0m0.729s
> user	0m0.653s
> sys	0m0.070s


2. `--database` version:
`
time perl vep -v -i examples/homo_sapiens_GRCh38.vcf --format vcf --database --fasta /home/enrico/.vep/fasta/Homo_sapiens.GRCh38.dna.toplevel.fa --vcf --offline --assembly GRCh38 --af_gnomad -o examples/forked.vcf --no_stats --hgvs --fork 4
`

> real	0m3.983s
> user	0m3.742s
> sys	0m0.248s

3. I profiled the run with `-d:NYTProf`, [here the results](https://github.com/cccnrc/nytprof-output-VEP100/tree/master/nytprof_output_files_VEP100)
`
perl -d:NYTProf vep -v -i examples/homo_sapiens_GRCh38.vcf --format vcf --cache --dir_cache /home/enrico/.vep --cache_version 100 --fasta /home/enrico/.vep/fasta/Homo_sapiens.GRCh38.dna.toplevel.fa --vcf --offline --assembly GRCh38 --af_gnomad -o examples/forked.vcf --no_stats --hgvs --fork 4 --force_overwrite
`

I was kind of surprised by the speed so I tried my original command again, and now it takes as short as it's supposed to do:
`
time perl vep -v -i examples/homo_sapiens_GRCh38.vcf --format vcf --cache --dir_cache /home/enrico/.vep --cache_version 100 --fasta /home/enrico/.vep/fasta/Homo_sapiens.GRCh38.dna.toplevel.fa --vcf --offline --assembly GRCh38 --af_gnomad -o examples/forked.vcf --no_stats --hgvs --fork 4 --force_overwrite
`
> real	0m3.903s
> user	0m3.571s
> sys	0m0.345s

No idea at all of why it should work this good right now. It may be due to the fact that I installed everything and didn't reboot first time I was using these commands? Why it should happen?

Really no idea

**Comment by aparton** [comment]
Hi @cccnrc,

I'm glad to hear that the issue has been resolved, however unsatisfying a spontaneous resolution is! Potentially there was something else running on the machine/node at the same time that was taking up resources, or an issue with the filesystem?

If the issue pops up again, I can take a closer look again if you send me NYTProf data.

Kind Regards,
Andrew

**Comment by cccnrc** [comment]
Great to know there is this chance with `NYTProf`, absolutely no idea of what could have been...

Thank you again @aparton !!!
